### PR TITLE
add /experimental/recommendations end point to image builder (HMS-3831)

### PR DESCRIPTION
**Pull request by mgold1234**
This commit introduces a new endpoint to the image builder API, enhancing its capabilities. The changes include:
1) Integration of a new client for the recommendations model
2) Addition of the '/experimental/recommendations' endpoint 
3) Inclusion of the client in configuration files
4) Implementation of a handler for 'recommendationPackage'
5) Introduction of new variables in template files
6) unit tests to check the added functionallity

### CODE REVIEW

**Comment by lzap** [comment]
LGTM, please update also `tools/update-spec.sh` and try it that it works. Feel free to update other clients as well but in separate commit.

**Comment by lzap** [comment]
So to give more context to other reviewers - the API that IB will call is currently internal only, it is a package recommendation AI model. Therefore we have decided to put the OpenAPI YAML manually for now and delete the URL from the document (internal URLs must not be public).

File: internal/v1/api.yaml
Comment (lzap) [inline internal/v1/api.yaml:670]: The original API is POST for some reason, in general this should be GET unless one modifies data or initiates some kind of big change (a job or something). Opinions? @ondrejbudai or @croissanne ?

File: internal/v1/api.yaml
Comment (lzap) [inline internal/v1/api.yaml:1332]: The original API returns an array in string weirdly: `"[package1, package2]"`

I asked the plan is to break it up into a proper JSON string, therefore, I suggest to expose it here correctly as a JSON array and for now parse the string and break it up manually until this is fixed upstream.

**Review by lzap — CHANGES_REQUESTED** [review]
As I sad in our yesterday conversation, the project will not build now as it is missing a method implementation. Please proceed with implementation, I left some other comments as separate comments. Looking good!

**Review by lzap — CHANGES_REQUESTED** [review]
Good changes, now you can implement the handler itself.

Also please make changes to clowder YAML to pass the URL into the container.

File: internal/v1/api.yaml
Comment (croissanne) [reply internal/v1/api.yaml:670]: Considering the request has a body, I think a POST is fine.

GET requests with bodies sometimes have weird results when they get cached etc… No strong opinion here but honestly as soon as a body is present it should not be a get imo

Comment (mgold1234) [reply internal/v1/api.yaml:670]: I will keep it "POST" for now

File: cmd/image-builder/main.go
Comment (lzap) [inline cmd/image-builder/main.go:102]: This does not look like a valid Go code to me.

File: internal/recommendations/client.go
Comment (lzap) [inline internal/recommendations/client.go:1]: File internal/recommendations/recommendations.v3.yml looks empty in this version of the patch.

File: internal/v1/handler_recommend_package.go
Comment (lzap) [inline internal/v1/handler_recommend_package.go:34]: Formatting.

File: internal/v1/api.go
Comment (lzap) [inline internal/v1/api.go:862]: The endpoint must change the component name, `UploadPackageRequest` is some sort of copy-paste leftover.

File: internal/v1/api.go
Comment (lzap) [inline internal/v1/api.go:807]: Why this is still named `Message`? From what I have hears, there are plans to break this response string into an array, I already see slice here so was this already done? Why this isn’t named `Packages` or something more relevant?

File: internal/v1/handler_recommend_package_test.go
Comment (lzap) [inline internal/v1/handler_recommend_package_test.go:52]: This endpoint should have never been POST, this should be GET and return just HTTP OK 200. This needs also to be fixed in the service first.

File: templates/image-builder.yml
Comment (lzap) [inline templates/image-builder.yml:277]: Please put some dummy value here, do not expose internal URLs, the real endpoint will be put into app-interface.

**Review by lzap — CHANGES_REQUESTED** [review]
I think the endpoint image builder is supposed to call must be fixed first:

* the operation should be GET not POST
* the request component name must be renamed
* the response should return properly named packages array

Then we should pull the updated YAML in and write the code.

File: templates/image-builder.yml
Comment (kingsleyzissou) [inline templates/image-builder.yml:277]: This URL looks incorrect

Comment (kingsleyzissou) [reply templates/image-builder.yml:277]: It's just an example. Ignore this.

File: internal/recommendations/client.go
Comment (croissanne) [inline internal/recommendations/client.go:143]: Needs to be client_credentials

File: internal/recommendations/client.go
Comment (croissanne) [inline internal/recommendations/client.go:141]: Can be dropped

File: cmd/image-builder/main.go
Comment (croissanne) [inline cmd/image-builder/main.go:100]: Can be dropped

**Comment by ondrejbudai** [comment]
May I ask you to change the title of this PR? It's currently very generic, which will look bad in the changelog. Thanks! :raised_hands:

File: internal/recommendations/client.go
Comment (croissanne) [inline internal/recommendations/client.go:115]:
```suggestion
	if resp.StatusCode == http.StatusForbidden || resp.StatusCode == http.StatusUnauthorized {
```

File: internal/recommendations/client.go
Comment (croissanne) [inline internal/recommendations/client.go:122]: Before redoing the request you need to reseek:
```suggestion
                _, err = body.Seek(0, io.SeekStart)
                if err != nil {
                        return nil, err
                }
		resp, err = rc.client.Do(req)
```

Comment (mgold1234) [reply internal/recommendations/client.go:122]: amazing, its work!! thank you so much

File: internal/v1/handler_recommend_package.go
Comment (lzap) [reply internal/v1/handler_recommend_package.go:52]: Yeah I added this to figure out why the endpoint returned an empty JSON to @mgold1234 probably not the best practice.

This all stems from the fact that the upstream service response is sort of too loose, there is no need the response should be optional, as you mention above.

File: internal/recommendations/recommendations.v3.json
Comment (mgold1234) [reply internal/recommendations/recommendations.v3.json:77]: and I think they both required, the default value of recommendedPackages is 3.

Comment (kingsleyzissou) [reply internal/recommendations/recommendations.v3.json:77]: Maybe we could make `packages` required and give `recommendedPackages` the default of 3 then :)

File: internal/v1/api.yaml
Comment (lzap) [inline internal/v1/api.yaml:1333]: Please add required flag to all fields for the same reason - the caller of this service does not need to fight with pointers, this must return something (e.g. an empty array).

**Review by lzap — CHANGES_REQUESTED** [review]
Looks good, since Sanne's PR is bigger I suggest this would wait until her's merged and I will ask you to rebase. You may squash into a single commit before rebasing to ease the pain of merging - there will be handful of conflicts unfortunately.

File: templates/image-builder.yml
Comment (lzap) [inline templates/image-builder.yml:285]: Merge problem

File: internal/v1/api.yaml
Comment (lzap) [reply internal/v1/api.yaml:1345]: Response object has `packages` element which is not required.

**Comment by lzap** [comment]
Tests red again, infra issue?

**Review by lzap — CHANGES_REQUESTED** [review]
Please move the client from `internal` package to `internal/clients/recommendations`.

File: internal/v1/handler_recommend_package_test.go
Comment (lzap) [inline internal/v1/handler_recommend_package_test.go:169]: Can you explain this test - what it does and why you have it?

File: internal/v1/handler_recommend_package_test.go
Comment (lzap) [inline internal/v1/handler_recommend_package_test.go:134]: Can you explain this test - what it does and why you have it?

File: internal/v1/handler_recommend_package_test.go
Comment (mgold1234) [reply internal/v1/handler_recommend_package_test.go:169]: I defined the server with 401 error
and the response after trying to post the request is StatusBadRequest

Comment (lzap) [reply internal/v1/handler_recommend_package_test.go:169]: What are you testing for? What is the use case?

The test is named ` TestRecommendPackage_Failed_with_wrong_url` so I assume it tests some use case when URL is wrong. That does not make sense to me, I see no wrong URL. Then it returns bad request.

You do not want to write tests for every single possible HTTP status code, in fact even those tests for 401/403 are unnecessary as long as token generation is extracted into separate reusable component with own tests.

What you want to cover with tests are success and failure scenarios or edge cases (empty response, empty string, `nil` etc). I would say that the positive test plus 401/403 tests are enough for this simple endpoint. I suggest to drop this test and the other one, unless you explain to me what is the purpose of this which I may be missing.

Comment (mgold1234) [reply internal/v1/handler_recommend_package_test.go]: agreee, removed those tests

File: internal/clients/recommendations/client.go
Comment (lzap) [inline internal/clients/recommendations/client.go:2]: Please move this comment into a separate file called `package.go` like it is done in https://github.com/osbuild/image-builder/pull/1128

But only if you get your tests red, in case you are green let’s merge this and I will update my PR :-)

File: internal/v1/api.yaml
Comment (lzap) [inline internal/v1/api.yaml:1342]: This field should be required in the response so the caller does not need to deal with pointers.

File: newrelic-infra.repo
Comment (lzap) [inline newrelic-infra.repo:6]: Uh is this relevant? Not to me.

**Review by lzap — CHANGES_REQUESTED** [review]
I think you accidentally added an unwanted file.

File: templates/image-builder.yml
Comment (croissanne) [inline templates/image-builder.yml:117]: recommendations-secrets are currently not present. We can either mark them as optional, or provision the secrets in the namespaces first.

Comment (mgold1234) [reply templates/image-builder.yml:117]: add it as optional

Comment (croissanne) [reply templates/image-builder.yml:117]: well you'd need to mark the client_id as optional as well.

**Comment by lzap** [comment]
We are missing one thing in clowder, please add RECOMMENDATIONS_TOKEN_URL with the same default value as composer:

```
  - name: COMPOSER_TOKEN_URL
    value: "https://xxxxx.redhat.com/xxxxxxxxx/openid-connect/token"
```

and make use of the variable in the `env` section.

**Review by lzap — DISMISSED** [review]
LGTM

**Comment by lzap** [comment]
```
namespace reserve: hit timeout error: Could not do 'waiting for namespace to be allocated to reservation 'bonfire-reservation-6b25ac8f'' at /var/lib/jenkins/workspace/osbuild-image-builder-pr-check@2/.bonfire_venv/lib64/python3.6/site-packages/bonfire/openshift.py:299 in time
14:52:04 ERROR: namespace reserve: hit timeout error: Could not do 'waiting for namespace to be allocated to reservation 'bonfire-reservation-6b25ac8f'' at /var/lib/jenkins/workspace/osbuild-image-builder-pr-check@2/.bonfire_venv/lib64/python3.6/site-packages/bonfire/openshift.py:299 in time
```

How can I restart just the failed test? Is that possible?

Triggered Rebuild in the CI UI: https://ci.ext.devshift.net/job/osbuild-image-builder-pr-check/1810/

**Comment by jrusz** [comment]
> ```
> namespace reserve: hit timeout error: Could not do 'waiting for namespace to be allocated to reservation 'bonfire-reservation-6b25ac8f'' at /var/lib/jenkins/workspace/osbuild-image-builder-pr-check@2/.bonfire_venv/lib64/python3.6/site-packages/bonfire/openshift.py:299 in time
> 14:52:04 ERROR: namespace reserve: hit timeout error: Could not do 'waiting for namespace to be allocated to reservation 'bonfire-reservation-6b25ac8f'' at /var/lib/jenkins/workspace/osbuild-image-builder-pr-check@2/.bonfire_venv/lib64/python3.6/site-packages/bonfire/openshift.py:299 in time
> ```
> 
> How can I restart just the failed test? Is that possible?
> 
> Triggered Rebuild in the CI UI: https://ci.ext.devshift.net/job/osbuild-image-builder-pr-check/1810/

Trigger from UI won't work. You have to put a `/retest` comment like I did above.

**Comment by lzap** [comment]
_ConsoleDotProxyError: Service Unavailable: status code: 503_

CRC proxy is down... jeeez.

**Review by lzap — APPROVED** [review]
LGTM and the only thing left is to pass the CI/CD pipeline.

🚀 GOOD LUCK WITH THAT 🚀

**Comment by jrusz** [comment]
I see the issue with pr_check, however I didn't expect that as it worked for me locally. Let me try again and figure out what's wrong...

**Comment by jrusz** [comment]
This should fix it https://github.com/osbuild/image-builder/pull/1139

**Comment by jrusz** [comment]
Rebased, we should see a pass now

**Comment by mgold1234** [comment]
@jrusz thank you so much for your help

**Review by kingsleyzissou — APPROVED** [review]

**Pull request merged**
